src/main/java/codeu/model/store/basic/IndexStore.java

```diff
@@ -66,13 +71,96 @@ public void addMapping (String word, Message message) {
   /**
     * Search function for a single word, returns a Set of messages from the Hashtable if the word is in it, otherwise return null
     */
-  public Set<Message> searchWord (String word) {
-    if (wordsHash.containsKey(word)) {
-      return wordsHash.get(word);
+  public List<Message> search (String searchTarget) {
+    if (searchTarget.contains("||")) {
+      return searchUnion(searchTarget);
+    }else if (searchTarget.contains("&&")) {
+      return searchIntersection(searchTarget);
+    }else if (wordsHash.containsKey(searchTarget)){
+      List<Message> messageList = new ArrayList<Message>(wordsHash.get(searchTarget));
+      return sortMessageList(messageList);
     }
     return null;
   }
 
+  /**
+    * Pre: the given string contains "||"
+    * returns the union of the elements in the given searchTarget
+    */
+  public List<Message> searchUnion (String searchTarget) {
+    // remove bad format || in searchtarget
+    if (searchTarget.indexOf("|") == 0) {
```

Comment:
Instead of all of this logic.  Is it functionally equivalent if you just ignore results in words that is empty string?  That would have the added benefit of handling cases such as "a || || b" and still returning a valid result.